cpp/src/barretenberg/proof_system/circuit_builder/circuit_simulator.hpp

```diff
@@ -0,0 +1,113 @@
+#pragma once
+#include "barretenberg/ecc/curves/bn254/fr.hpp"
+#include "barretenberg/proof_system/arithmetization/gate_data.hpp"
+#include "barretenberg/proof_system/types/merkle_hash_type.hpp"
+#include "barretenberg/proof_system/types/pedersen_commitment_type.hpp"
+
+namespace proof_system {
+
+class CircuitSimulatorBN254 {
+  public:
+    using FF = barretenberg::fr;                                                 // IOU templating
+    static constexpr merkle::HashType merkle_hash_type = merkle::HashType::NONE; // UGH
+    static constexpr pedersen::CommitmentType commitment_type = pedersen::CommitmentType::NONE;
+
+    bool contains_recursive_proof = false;
+    static constexpr size_t UINT_LOG2_BASE = 2; // WORKTODO: 6 for Ultra
+
+    const uint32_t zero_idx = 0;
+
+    // uint32_t add_variable([[maybe_unused]]const FF& in){
+    //   return 0; // WORKTODO: return part of `in` for debugging purposes?
+    // }
+
+    inline uint32_t add_variable([[maybe_unused]] const barretenberg::fr index) const { return 1028; }
```

Comment:
1028 = 0x404, recognizable when printing if something went wrong in the simulation.